### ISSUE TITLE
Custom User backend

### ISSUE THREAD

**Issue by TaaviE** [jwag956/flask-security#139]
I'm currently trying to get multi-email, better audit logs and password history support into the original Flask-Security but that has proven non-trivial. Are there any plans to make it easier in this fork or even integrate the functionality?

Just that people reading this would understand the feature request better: 
* I'm working on multi-email support because some people want to log in with multiple e-mails or log in with one and get emails to another, the current db model is unsuitable for that. 
* Password history support is nice to be able to remind people that they've changed passwords or to enforce no password reuse when compromise is detected. 
* Audit logs are somewhat self-explanatory - to track changes made to accounts and logins so people know what's going on with their logins and to provide warnings if needed. The current rudimentary configuration option and data model is just terribly unsuitable for that.

**Comment by TaaviE** [comment]
I'd like to get this functionality in the main branch so when I'd create a PR would it be okay if I implemented a getter for the email and password field in the User object so that it would be nicer to implement such functions in alternate ways?

**Comment by jwag956** [comment]
I am in the process of writing a longer response - some of the things you propose make sense for inclusion in the master branch - some I think are functionality we should support via overriding and API extensions.

**Comment by jwag956** [comment]
Lots to think about. For any feature concepts like these there is a 3 step process to evaluate:

1) is this a security feature consistent with what flask-security is trying to provide
2) does FS already have hooks that would make this straight forward to implement (without having to delve into the innards of FS
3) is it of general value that it is worth adding to the base FS code - thereby requiring long-term support.

Here are some of my initial thoughts - I welcome a discussion.

Multi-email support - Being able to use different fields in the User model to log in is supported - there have been bugs but I believe my fork has already fixed them. While it is common to be able to log into websites with either an email or username - I haven't seen many sites that do much beyond that. SECURITY_USER_IDENTITY_ATTRIBUTES is designed to do just that. You could easily add any number of fields into the User model (e.g. secondary_email) and decide which ones are allowed to be used to log in.

passwords - I have an existing issue aimed at making it easy to plug in password validators. There are a few ways to do this - one would be to simply add a 'validate_password' method as part of the UserMixin - just as verify_and_update_password was a few months ago. That would give application writers complete control - they own the user model and can do whatever checks etc they want. I don't think FS should do much in this regard - there are too many different ways apps might want to limit passwords.

Audit logs - my first reaction is that yes - the UserModel isn't good at all for that - you would want some other table or even dump things like that into a document store. I haven't looked that closely - but it sure seems like the currently supported signals give the application access to all those changes such that it could then dump whatever info it wants into a table. Is there something specific missing?

**Comment by TaaviE** [comment]
> You could easily add any number of fields into the User model (e.g. secondary_email) and decide which ones are allowed to be used to log in.

Having multiple columns for multiple e-mails (or user identity attributes in general) is very rigid and management requires quite a bit of effort. I think that a many-to-one relationship would be more suitable, that there's a `emails` table that specifies to whom the email belongs to and easy management of if it is primary, is it verified and so on. It would mean less hassle than say juggling empty e-mail columns and trying to enforce some constraints on those.

It is not however a strictly security feature, however is a big part of user management flow FS currently provides. FS does not have any hooks that I've seen that would allow overriding how it gets the e-mail of an user, seems to be `user_object.email` mostly. If the goal is batteries-included I think the batteries should be more standard (`AA`) rather than something rarer (`4LR44`).

>  I don't think FS should do much in this regard - there are too many different ways apps might want to limit passwords.

This sounds good. However 

> but it sure seems like the currently supported signals give the application access to all those changes such that it could then dump whatever info it wants into a table. Is there something specific missing?

In theory yes, it's doable with signals, but I think it could be done nicer out-of-box especially if there's `SECURITY_TRACKABLE` that tries to offer some of the functionality. The nice way would be to create another table to log all changes instead of updating some user columns and have that out-of-box.


I'm thinking the easiest way for me to get the functionality I want is to keep the current User object but create a few new tables I've mentioned and a few stored procedures that deal with keeping e-mail info latest and logging password (etc.) changes. However I doubt I'm the only one who would want these features.

**Comment by jwag956** [comment]
OK - so we have 3 different possible enhancements. Lets make sure we split those up in terms of PRs and discussion. I think we are agreed that for password policy we can use #99 - which I think is pretty easy and I do think many applications would want to use.

So possibly we should split the other 2 out - 'enchanced user model' and 'auditability'.
For auditability - I would consider this a new 'feature' which, when enabled would require a new table that was more 'audit/append-only' style. I can see that as useful for some folks.

For 'enchanced user model' I am not really seeing the a larger use case than yours. However, making the small change that wouldn't REQUIRE that the email be physically present in the UserModel by providing a getter seems fine.

**Comment by TaaviE** [comment]
In the case of e-mails and passwords it seems to me that the fastest and cleanest solution would be leaving UserMixin the same (no getters) but creating the additional tables for emails and passwords and then have stored procedures that update the User rows with current password and e-mail. In addition to that the password updating function would just have to add a new row to passwords and invalidate old one. There's only one problem though, stored procedures would have to exist per-db. I will write this for PostgreSQL but would there be interest in having just that upstream?

With just audit logging, there's only need to rewrite the TRACKABLE features into into inserting rows into some new table, that I can implement and submit a PR.

**Comment by jwag956** [comment]
So as I mentioned - I am not  convinced that there is need in the base FS for that level of complexity - and I certainly don't want things like stored procedures etc in the basic usage model for FS.
Furthermore, which DB, which ORM is likely to vary a huge amount - it is already hard enough to test (you might have noticed that I added some real DB testing as part of CI).
I think it's important FS makes as little requirements as possible in terms of sessions, transactions etc. That doesn't mean I don't want to provide some 'best practice' - such as what I did in the models/fsqla.py) - just not part of the basic 'getting started'.

As for audit - we can discuss more - but my current thoughts are that we should add a new feature rather than change the current one. Trackable gives current status - not audit. Adding auditing could be nice - I don't think most apps (including mine!) have any interest in doing auditing.

**Comment by TaaviE** [comment]
> but my current thoughts are that we should add a new feature rather than change the current one.

That sounds reasonable, I'll try a few approaches.

**Comment by TaaviE** [comment]
I tried out using `@property` getters and setters to redirect data flows to keep an audit log and to separate users table and emails, the result was ugly in the code sense (`@property` is not very Pythonic), in the end I also discovered it could cause data inconsistencies very easily when any bugs or race conditions happen. 

As a second approach I wrote a few stored procedures for PostgreSQL that provided me multiple e-mails per user and it also audit logging. It also has the benefit of not requiring any changes to Flask-Security and Flask-Security-Too, both can do what they want without any additional (ugly) code - the trigger just intercepts and keeps track of all the possible changes. It isn't super simple to turn off but it really seems like the cleanest solution. If anyone else is interested then I'll push it to https://github.com/TaaviE/Loosindus at some point, needs a bit of polish UI side.
